### PR TITLE
Implement SetCombinationSuppliers & RemoveAllCombinationSuppliers commands

### PR DESCRIPTION
**Pull request by zuk3975**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Implement SetCombinationSuppliersCommand & refacto existing SetProductSuppliersHandler. It also includes refactoring of related ProductSupplierDTO's and repository method to allow segregating ProductSuppliers by combination id. Covers with behat tests.
| Type?             | refacto
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | part of #23002 
| How to test?      | CI :heavy_check_mark: 
| Possible impacts? | Please indicate what parts of the software we need to check to make sure everything is alright.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/23284)

### CODE REVIEW

File: src/Core/Domain/Product/Combination/Command/SetCombinationSuppliersCommand.php
Comment (eternoendless) [inline src/Core/Domain/Product/Combination/Command/SetCombinationSuppliersCommand.php:58]: Watch out, remember [the principles](https://devdocs.prestashop.com/1.7/development/architecture/domain/cqrs/#command-and-commandhandler-principles):

> 3. A Command receives only primitive types on input (int, float, string, bool and array).

If you send a collection of classes, then the command is harder to serialize.

Comment (eternoendless) [reply src/Core/Domain/Product/Combination/Command/SetCombinationSuppliersCommand.php:58]: This is debatable though...

**Comment by zuk3975** [comment]
TODO:
* modify existing "RemoveAllAssociatedProductSuppliersCommand" to avoid removing combinational product suppliers and add a "RemoveAllAssociatedCombinationSuppliersCommand" for that.
* Add couple more test cases

File: src/Core/Domain/Product/Combination/Command/SetCombinationSuppliersCommand.php
Comment (zuk3975) [reply src/Core/Domain/Product/Combination/Command/SetCombinationSuppliersCommand.php:58]: yep that was a mistake, I've homogenized with SetProductSuppliersCommand. It accepts an array and transforms it to object in constructor.

File: src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php
Comment (zuk3975) [inline src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php]: Wdyt about this?
I know we prefer composition over inheritance, but in this case I dont know where to put this kind of service. (it could be some sort of DataProvider, but it is actually specific to these queries only).
I could return SuppliersInfo[] straight inside repository, but I don't want to couple the repository method with the usecase either.

File: tests/Integration/Behaviour/Features/Scenario/Product/Combination/update_combination_suppliers.feature
Comment (Progi1984) [inline tests/Integration/Behaviour/Features/Scenario/Product/Combination/update_combination_suppliers.feature:116]: Could you check that all another combinations have not  any suppliers assigned

File: src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php
Comment (Progi1984) [inline src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php]:
```suggestion
     * @return array<int, ProductSupplierInfo>
```

File: src/Adapter/Product/Update/ProductSupplierUpdater.php
Comment (Progi1984) [inline src/Adapter/Product/Update/ProductSupplierUpdater.php]:
```suggestion
        return array_map($existingProductSuppliers, function(array $currentSupplier) {
            return new ProductSupplierId((int) $currentSupplier['id_product_supplier']);
        });
    }
```

**Review by Progi1984 — COMMENTED** [review]
Some feedbacks

File: src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php
Comment (zuk3975) [reply src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php]: although I think its quite clear as it is, I'll update it and then this means its a new convention? - we use psalm notations everywhere (and not `Object[]`) :thinking:

Comment (jolelievre) [reply src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php]: Yeah I'm not sure this notation brings out much values 🤷‍♂️
For more advanced array format it can be very valuable but in this case it's just a simple array of objects What's the interest of this despite making PHPStan happy maybe? If it's just for PHPStan I don't think we should let him dictate what we need to do if it doesn't bring any value for us
Don't let robots control you!! 😜

File: src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php
Comment (jolelievre) [reply src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php]: Personally I'm fine with it is just a function that formats raw data from Repository into DTO objects adapted to the command handler, so it makes sense that it is handed by the handler and not the Repository itself

Besides Product and Combination handlers are very similar, so it makes sens to allow them to share this abstract class, we correctly split the Domain commands which was the most important Sharing an abstract class doesn't couple them since we can always remove this inheritance if needed without causing any breaking change

This is just convenient to avoid duplicating code, I would have preferred a trait to handle this but since they are not appreciated among the maintainers I guess this abstract class is a good compromise ^^

File: src/Adapter/Product/Combination/QueryHandler/GetCombinationSupplierOptionsHandler.php
Comment (jolelievre) [inline src/Adapter/Product/Combination/QueryHandler/GetCombinationSupplierOptionsHandler.php]: I get the impression that this `CombinationSupplierOptions` doesn't have much interest since it only carries a list a `ProductSupplierInfo`

I think the handler interface could be simplified by just returning an array of `CombinationSupplierOptions` objects, for product the `ProductSupplierOptions` additional layer had meaning since it also had the default supplier ID But here not so much 🤔

File: src/Adapter/Product/Combination/CommandHandler/SetCombinationSuppliersHandler.php
Comment (jolelievre) [inline src/Adapter/Product/Combination/CommandHandler/SetCombinationSuppliersHandler.php]:
```suggestion
        $productSupplier = new ProductSupplier($id);
```

Creation and update shouldn't be handled the same way, when the ID is present we should use it to fetch the existing data, this brings two interest:
- we don't manage it yet but for partial updates we need to get existing object and update only required fields, here all fields are overwritten but some of them could be null in the future
- using the ID should allow to check that the ProductSupplier does exist in database, in that spirit maybe we should even rely on `ProductSupplierRepository` to get the object while checking its existence

File: src/Adapter/Product/CommandHandler/SetProductSuppliersHandler.php
Comment (jolelievre) [inline src/Adapter/Product/CommandHandler/SetProductSuppliersHandler.php]: Same suggestion as for Combination the ProductSupplier should be fetched by ID

File: src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php
Comment (jolelievre) [reply src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php]: Should `buildEntityFromDTO` also be in this abstract class?

File: src/Adapter/Product/Update/ProductSupplierUpdater.php
Comment (jolelievre) [inline src/Adapter/Product/Update/ProductSupplierUpdater.php:91]: @Progi1984 was this really necessary? (see more details in this comment https://github.com/PrestaShop/PrestaShop/pull/23284#discussion_r580240536)

File: src/Adapter/Product/Update/ProductSupplierUpdater.php
Comment (jolelievre) [inline src/Adapter/Product/Update/ProductSupplierUpdater.php]:
```suggestion
    private function persistSuppliers(ProductId $productId, array $productSuppliers, ?CombinationId $combinationId = null): void
```

Comment (jolelievre) [reply src/Adapter/Product/Update/ProductSupplierUpdater.php]: Suggestion, but `persist` alone seems too generic for this private function (even though we are in a SuppliersRepository)

File: src/Adapter/Product/Update/ProductSupplierUpdater.php
Comment (jolelievre) [inline src/Adapter/Product/Update/ProductSupplierUpdater.php]: Isn't there a thing like product with combination has no default supplier?

File: src/Core/Domain/Product/Supplier/Command/RemoveAllAssociatedProductSuppliersCommand.php
Comment (jolelievre) [inline src/Core/Domain/Product/Supplier/Command/RemoveAllAssociatedProductSuppliersCommand.php]:
```suggestion
 * @see RemoveAllAssociatedProductSuppliersHandlerInterface
```

I guess?

File: src/Adapter/Product/Combination/CommandHandler/RemoveAllAssociatedCombinationSuppliersHandler.php
Comment (jolelievre) [inline src/Adapter/Product/Combination/CommandHandler/RemoveAllAssociatedCombinationSuppliersHandler.php]: As we discussed all combination handlers should first check that the product type (has combinations) and throw an exception if it's not the case
And vice versa, the regular product handlers should check it's not a combination type

File: tests/Integration/Behaviour/Features/Context/Domain/Product/Combination/UpdateCombinationSuppliersFeatureContext.php
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/Product/Combination/UpdateCombinationSuppliersFeatureContext.php:56]: Shouldn't we throw an exception of the reference doesn't exist? It means the reference is wrongly written or something went wrong
And since `get` already handles this exception I guess we could simply remove this check no?

Comment (jolelievre) [reply tests/Integration/Behaviour/Features/Context/Domain/Product/Combination/UpdateCombinationSuppliersFeatureContext.php:56]: If you need to handle creation/update switch you'd rather rely on `if (!empty($row['reference']))`

Comment (jolelievre) [reply tests/Integration/Behaviour/Features/Context/Domain/Product/Combination/UpdateCombinationSuppliersFeatureContext.php:56]: Oh I think I got it, on creation the reference is not accessible since it was never created But it is used later to save its ID That makes sense!

File: tests/Integration/Behaviour/Features/Context/Domain/Product/Combination/UpdateCombinationSuppliersFeatureContext.php
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/Product/Combination/UpdateCombinationSuppliersFeatureContext.php:84]: I'm a bit afraid this could cause bugs if the returned array and the input one are not exactly in the same order, then the IDs would not be associated to the right reference (not sure if it can be handled better though)

File: src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php
Comment (zuk3975) [reply src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php]: php stan is ok with previous notation.

File: src/Adapter/Product/Combination/QueryHandler/GetCombinationSupplierOptionsHandler.php
Comment (zuk3975) [reply src/Adapter/Product/Combination/QueryHandler/GetCombinationSupplierOptionsHandler.php]: yeah, I had same idea too, but thought this way we leave some space to easily add anything else if we need to that query (as we did need default supplier in standard product).
But I suppose it might be the [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) :smile: so i'll simplify it

File: src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php
Comment (zuk3975) [reply src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php]: no, this is for query only, but if its alright, then it would make sense putting both of them in more abstract "AbstractProductSupplierHandler".

File: tests/Integration/Behaviour/Features/Context/Domain/Product/Combination/UpdateCombinationSuppliersFeatureContext.php
Comment (zuk3975) [reply tests/Integration/Behaviour/Features/Context/Domain/Product/Combination/UpdateCombinationSuppliersFeatureContext.php:84]: Expected data should be order-sensitive, I think its alright :man_shrugging: (it would be strange if in any order we would get green tests, like we couldnt test sorting if we wanted too)

File: src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php
Comment (zuk3975) [reply src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php]: but we wanted to get rid of these abstract classes to promote the composition over inheritence, thats why im hesitating with this one.

File: src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php
Comment (zuk3975) [reply src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php]: so, @Progi1984 , @jolelievre , @matks could we agree on which one to use when we have simple array?

File: src/Adapter/Product/Update/ProductSupplierUpdater.php
Comment (zuk3975) [reply src/Adapter/Product/Update/ProductSupplierUpdater.php]: we are in ProductSupplierUpdater. To me It sounds exactly what it does :smile: 
`persistSuppliers` doesn't make sense because we are persisting ProductSuppliers :stuck_out_tongue: 
But I suppose you ment persistProductSuppliers. That I can do, altough its abit redundant.

File: src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php
Comment (zuk3975) [reply src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php]: btw I think it would make sense to use `ProductSupplierInfo[]` notations where its simple array (index starts at 0) and maybe array<int, ProductSupplierInfo> in case array keys are ids or some other custom numbers (the idea would be to point out that its not ordinary php array starting from 0).

File: src/Adapter/Product/Combination/CommandHandler/RemoveAllAssociatedCombinationSuppliersHandler.php
Comment (zuk3975) [reply src/Adapter/Product/Combination/CommandHandler/RemoveAllAssociatedCombinationSuppliersHandler.php]: hmm In this case (combination handlers) the first thing that it does is tries to load the combination, so if there is none, it will throw the exception. (because we don't even provide ProductId in this command).
in normal handlers additional check should help tho.

File: src/Adapter/Product/Update/ProductSupplierUpdater.php
Comment (zuk3975) [reply src/Adapter/Product/Update/ProductSupplierUpdater.php]: yes there is. Product with combination should not have a default supplier

File: src/Adapter/Product/Update/ProductSupplierUpdater.php
Comment (jolelievre) [inline src/Adapter/Product/Update/ProductSupplierUpdater.php]: I think it deserves a dedicated exception something like `InvalidProductType` or similar (same in `setProductSuppliers`)

File: src/Adapter/Product/Update/ProductSupplierUpdater.php
Comment (jolelievre) [inline src/Adapter/Product/Update/ProductSupplierUpdater.php:168]: Can you add a check similar to `removeAllForProduct` and `setProductSuppliers` If the product is not a combination type throw an exception (same in `setCombinationSuppliers`)

File: src/Core/Domain/Product/Combination/Command/SetCombinationSuppliersCommand.php
Comment (jolelievre) [inline src/Core/Domain/Product/Combination/Command/SetCombinationSuppliersCommand.php]:
```suggestion
     * @return ProductSupplier[]
```

File: src/Core/Domain/Product/Combination/Command/SetCombinationSuppliersCommand.php
Comment (jolelievre) [inline src/Core/Domain/Product/Combination/Command/SetCombinationSuppliersCommand.php:94]: Not to handle in this PR but I personally find the class name `ProductSupplier` a bit confusing, we kept the legacy name but maybe we could have improved it (like we did with VirtualProduct)
I thought `SupplierReference` might be clearer, but we can discuss this later

File: tests/Integration/Behaviour/Features/Context/Domain/Product/Combination/UpdateCombinationSuppliersFeatureContext.php
Comment (jolelievre) [reply tests/Integration/Behaviour/Features/Context/Domain/Product/Combination/UpdateCombinationSuppliersFeatureContext.php:84]: yes, I'm a bit afraid because I remember having some issues in another behat test where references were not saved  correctly, it was not detected at first but it ended in false positive later But I can't remember where it was, and anyway it seems ok and it works so it's fine ^^

File: src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php
Comment (Progi1984) [reply src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php]: Please always use Psalm notations, it will be required in PHPStan level 6.

File: tests/Integration/Behaviour/Features/Scenario/Product/Combination/update_combination_suppliers.feature
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Product/Combination/update_combination_suppliers.feature:127]: Can you add some tests for invalid types (once a dedicated exception will be created) For both scenarios:
- forbidden combination action on regular product (I guess it will be in the other feature file)
- forbidden regular action on combination product

File: src/Adapter/Product/Update/ProductSupplierUpdater.php
Comment (jolelievre) [inline src/Adapter/Product/Update/ProductSupplierUpdater.php:129]: If I remember right we discussed about an issue, when you switch from simple to combination product There may be some ProductSupplier left with combinationId equal to zero Shouldn't they be cleaned here?

(same question in `setProductSuppliers` maybe we should remove all former ProductSupplier associated to former combinations?)

File: src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php
Comment (zuk3975) [reply src/Adapter/Product/QueryHandler/AbstractSupplierOptionsHandler.php]: oh, well okay then

File: src/Core/Domain/Product/Combination/Command/SetCombinationSuppliersCommand.php
Comment (zuk3975) [reply src/Core/Domain/Product/Combination/Command/SetCombinationSuppliersCommand.php:94]: ProductSupplier is ok IMO, it defines the relation purpose pretty well. (SupplierReference might be even more confusing). Maybe the problem is that this dto is very generic and its not that easy to understand what its used for. Anyway I agree there is room for improvement, but not in this PR.

File: src/Adapter/Product/Update/ProductSupplierUpdater.php
Comment (jolelievre) [inline src/Adapter/Product/Update/ProductSupplierUpdater.php]: I think it may be safer to use `$product->hasCombinations()` just in case `cache_default_attribute` isn't consistent I know it implies an extra query but it seems safer

I'm not blocking the PR for this though, so I leave the decision to you 😉

File: src/Adapter/Product/Update/ProductSupplierUpdater.php
Comment (jolelievre) [inline src/Adapter/Product/Update/ProductSupplierUpdater.php]:
```suggestion
        if ($product->hasCombinations()) {
```

Same suggestion (optional)

File: src/Adapter/Product/Update/ProductSupplierUpdater.php
Comment (jolelievre) [inline src/Adapter/Product/Update/ProductSupplierUpdater.php]:
```suggestion
    private function getProductSupplierIds(ProductId $productId, ?CombinationId $combinationId = null): array
```

**Review by jolelievre — DISMISSED** [review]
Couple suggestions but nothing blocking

File: tests/Integration/Behaviour/Features/Context/Domain/Product/UpdateProductSuppliersFeatureContext.php
Comment (Progi1984) [inline tests/Integration/Behaviour/Features/Context/Domain/Product/UpdateProductSuppliersFeatureContext.php]:
```suggestion
    public function assertLastErrorInvalidProductType(): void
```

File: src/Core/Domain/Product/Combination/Command/SetCombinationSuppliersCommand.php
Comment (Progi1984) [reply src/Core/Domain/Product/Combination/Command/SetCombinationSuppliersCommand.php]:
```suggestion
     * @return array<int, ProductSupplier>
```

File: src/Core/Domain/Product/Combination/Command/SetCombinationSuppliersCommand.php
Comment (Progi1984) [inline src/Core/Domain/Product/Combination/Command/SetCombinationSuppliersCommand.php]:
```suggestion
     * @var array<int, ProductSupplier>
```

**Review by Progi1984 — COMMENTED** [review]
Some feedbacks

File: src/Core/Domain/Product/Exception/InvalidProductTypeException.php
Comment (matks) [inline src/Core/Domain/Product/Exception/InvalidProductTypeException.php:34]: 👍

File: tests/Integration/Behaviour/Features/Context/Domain/Product/Combination/UpdateCombinationSuppliersFeatureContext.php
Comment (matks) [inline tests/Integration/Behaviour/Features/Context/Domain/Product/Combination/UpdateCombinationSuppliersFeatureContext.php:44]:
```suggestion
     * @When I mark that combination ":combinationReference" is provided by the following suppliers:
```
Should we try to make our behat tests more "merchant oriented"?

File: tests/Integration/Behaviour/Features/Context/Domain/Product/Combination/UpdateCombinationSuppliersFeatureContext.php
Comment (matks) [inline tests/Integration/Behaviour/Features/Context/Domain/Product/Combination/UpdateCombinationSuppliersFeatureContext.php:90]:
```suggestion
     * @When I mark combination ":combinationReference" is not provided by suppliers anymore
```
Following previous idea?

File: tests/Integration/Behaviour/Features/Context/Domain/Product/Combination/UpdateCombinationSuppliersFeatureContext.php
Comment (matks) [inline tests/Integration/Behaviour/Features/Context/Domain/Product/Combination/UpdateCombinationSuppliersFeatureContext.php:106]:
```suggestion
     * @Then combination ":combinationReference" should be provided/produced by following suppliers:
```

**Review by matks — APPROVED** [review]
Code is OK for me. I only added not-blocking suggestions about behat wording, thinking about product / merchant oriented statements

File: tests/Integration/Behaviour/Features/Context/Domain/Product/Combination/UpdateCombinationSuppliersFeatureContext.php
Comment (zuk3975) [reply tests/Integration/Behaviour/Features/Context/Domain/Product/Combination/UpdateCombinationSuppliersFeatureContext.php:44]: Hmm Im not against it, but Im not the one to decide, need more opinions :stuck_out_tongue:

**Review by jolelievre — APPROVED** [review]

**Comment by Flo340** [comment]
Hello
I can't apply the fix on **prestashop 1.7.8.5**
Do you know how to do it
Thanks a lot for your help

**Pull request merged**
